MDANSE/Src/MDANSE/Chemistry/GroupingTool.py

```diff
@@ -0,0 +1,415 @@
+#    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from __future__ import annotations
+from typing import Optional, Any
+from enum import Enum, auto
+from collections import Counter, defaultdict
+from multiprocessing import Lock
+
+import numpy as np
+
+from MDANSE.MLogging import LOG
+from MDANSE.Framework.OutputVariables.IOutputVariable import OutputData
+from MDANSE.Chemistry.ChemicalSystem import ChemicalSystem
+
+
+class GroupingLevel(Enum):
+    ATOM = auto()
+    MOL_EACH = auto()
+    MOL_AVERAGE = auto()
+
+
+GROUPING_LABELS = {
+    "atom": GroupingLevel.ATOM,
+    "individual molecules": GroupingLevel.MOL_EACH,
+    "average over molecules": GroupingLevel.MOL_AVERAGE,
+}
+
+BACKUP_DATA_PARAMETERS = {
+    "axis": "index",
+    "units": "au",
+    "main_result": True,
+    "partial_result": True,
+    "dtype": np.float64,
+}
+
+
+class GroupingTool:
+    """Handles the analysis results and assigns the
+    correct results to different datasets based on the grouping
+    setting"""
+
+    def __init__(self, chemical_system: ChemicalSystem, output_data: OutputData):
+        """Set grouper up using trajectory topology.
+
+        Parameters
+        ----------
+        chemical_system : ChemicalSystem
+            information about atoms and molecules in the system
+        output_data : OutputData
+            structure into which to write the results of the calculation
+
+        """
+        self._cs = chemical_system
+        self._output_data = output_data
+        self._grouping = None
+        self._data_parameters = {}
+        self._weight_dictionary = {}
+        self._mass_dictionary = {}
+        self._molecule_mass = defaultdict(float)
+        self._current_selection = set()
+        self._indices_per_data_key = {}
+        self._mandatory_keys = ["output_type", "dimensions"]
+        self._extra_keys = ["axis", "units", "main_result", "partial_result", "dtype"]
+        self._plain_datasets = set()
+        self._weighted_datasets = set()
+        self._molecule_datasets = set()
+        self._mutex = Lock()
+        self._debug_counter = Counter()
+
+    def set_grouping(self, text_key: str):
+        """Choose the level of grouping to use.
+
+        This has to be configured before creating output data sets.
+
+        Parameters
+        ----------
+        text_key : str
+            name of the grouping level in GROUPING_LABELS
+
+        """
+        self._grouping = GROUPING_LABELS.get(text_key, GroupingLevel.ATOM)
+
+    def set_weight_dictionary(self, new_weights: dict[tuple[str], float]):
+        """Store the dictionary with weights per atom type.
+
+        Parameters
+        ----------
+        new_weights : dict[tuple[str],float]
+            dictionary of {atom_types: scaling_factor} pairs
+
+        """
+        self._weight_dictionary = {
+            "".join([*key]): value for key, value in new_weights.items()
```

Comment:
```suggestion
            "".join(key): value for key, value in new_weights.items()
```
If key is a sequence this will do.